### PR TITLE
[WIP][Bugfix] Fix flaky topk

### PR DESCRIPTION
**Pull request by sxjscience**
## Description ##
This PR fixes the flaky topk test reported in https://github.com/apache/incubator-mxnet/pull/12358 , https://github.com/apache/incubator-mxnet/issues/12310. The previous bug is caused by not manually setting the dtype of mx.ndarrays when constructing them  from numpy ndarrays (related issue: https://github.com/apache/incubator-mxnet/issues/12268). After reimplementing the `IndexFill` function, the test passes now.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] fix topk and its testing codes.

## Comments ##
- we really need to fix https://github.com/apache/incubator-mxnet/issues/12268.

### CODE REVIEW

**Comment by sxjscience** [comment]
@ankkhedia

**Comment by marcoabreu** [comment]
Any reason you are removing int16 support?

**Comment by sxjscience** [comment]
@macroabreu That's because the ndarray in MXNet does not support int16. So I removed it.

**Comment by sxjscience** [comment]
Find that it still raises a CUDA error. Need more time to fix it.
________________________________

**Comment by ankkhedia** [comment]
@sxjscience ping! 

Did you get some chance to look into the failure? This might be required for fixing the flaky test I referenced.

**Comment by kalyc** [comment]
@sxjscience could you please update the issue? Resolving a flaky test - https://github.com/apache/incubator-mxnet/pull/12358 is blocked because of this.

**Comment by sxjscience** [comment]
I’ll look into this again. Sorry for the delay.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by kalyc** [comment]
@mxnet-label-bot[pr-awaiting-response]

File: src/operator/tensor/ordering_op-inl.h [hunk 1]
```diff
@@ -455,8 +457,7 @@ void TopKImpl(const RunContext &ctx,
   // Cast `ret_indices` from int to real_t could introduce conversion error when the element_num
   // is large enough.
   if (param.ret_typ == topk_enum::kReturnMask) {
-    Tensor<xpu, 2, DType> ret_mask =
-      ret[0].get_with_shape<xpu, 2, DType>(Shape2(ret[0].Size(), 1), s);
+    Tensor<xpu, 1, DType> ret_mask = ret[0].FlatTo1D<xpu, DType>(s);
     ret_mask = scalar<DType>(0);
```
Comment (sxjscience) [inline src/operator/tensor/ordering_op-inl.h]: Now it raises a really weird "CUDA Misaligned Memory Error". I currently having no idea what triggers it. Actually it happens when we initialize the ret_mask to all zero.

Comment (lupesko) [reply src/operator/tensor/ordering_op-inl.h]: @anirudh2290 @azai91 @apeforest @samskalicky @eric-haibin-lin - maybe one of you guys can help?

Comment (vandanavk) [reply src/operator/tensor/ordering_op-inl.h]: @sxjscience were you able to resolve the error?

**Comment by lebeg** [comment]
@sxjscience once you've fixed this, could you submit a PR to 1.3.x as well? Currently the branch build is broken due to the `test_operator_gpu.test_order` test: http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/v1.3.x/40/pipeline

**Pull request closed without merging**